File: zig/3742.zig
const std = @import("std");

fn solution(comptime m: usize, comptime n: usize, allocator: std.mem.Allocator, grid: [m][n]i32, k: usize) !i32 {
    var table = try allocator.alloc([][]i32, m);
    for (0..m) |i| {
        table[i] = try allocator.alloc([]i32, n);

        for (0..n) |j| {
            table[i][j] = try allocator.alloc(i32, k + 1);

            for (0..k + 1) |a| {
                table[i][j][a] = -1;
            }
        }
    }

    defer {
        for (0..m) |i| {
            for (0..n) |j| {
                allocator.free(table[i][j]);
            }

            allocator.free(table[i]);
        }

        allocator.free(table);
    }

    var sc: usize = 0;
    if (grid[0][0] > 0) {
        sc = 1;
    }

    if (sc <= k) {
        table[0][0][sc] = grid[0][0];
    }

    for (0..m) |i| {
        for (0..n) |j| {
            var cost: usize = 0;
            if (grid[i][j] > 0) {
                cost = 1;
            }

            if (i > 0) {
                for (0..k + 1) |a| {
                    if (a + cost <= k and table[i - 1][j][a] > -1) {
                        table[i][j][a + cost] = @max(table[i][j][a + cost], table[i - 1][j][a] + grid[i][j]);
                    }
                }
            }

            if (j > 0) {
                for (0..k + 1) |a| {
                    if (a + cost <= k and table[i][j - 1][a] > -1) {
                        table[i][j][a + cost] = @max(table[i][j][a + cost], table[i][j - 1][a] + grid[i][j]);
                    }
                }
            }
        }
    }

    var best: i32 = -1;
    const target = table[m - 1][n - 1];

    for (0..k + 1) |i| {
        best = @max(best, target[i]);
    }

    return best;
}

const expect = std.testing.expectEqual;

test "test case #1" {
    const allocator = std.testing.allocator;

    const m = 2;
    const n = 2;

    const grid = [m][n]i32{
        .{ 0, 1 },
        .{ 2, 0 },
    };
    const k = 1;

    const expected = 2;
    const actual = try solution(m, n, allocator, grid, k);

    try expect(expected, actual);
}

test "test case #2" {
    const allocator = std.testing.allocator;

    const m = 2;
    const n = 2;

    const grid = [m][n]i32{
        .{ 0, 1 },
        .{ 1, 2 },
    };
    const k = 1;

    const expected = -1;
    const actual = try solution(m, n, allocator, grid, k);

    try expect(expected, actual);
}

File: go/3742.go
package main

func greater(a, b int) int {
	if a > b {
		return a
	}

	return b
}

func maxPathScore(grid [][]int, k int) int {
	table := make([][][]int, len(grid))
	for i := 0; i < len(grid); i++ {
		table[i] = make([][]int, len(grid[i]))
		for j := range table[i] {
			table[i][j] = make([]int, k+1)

			for c := 0; c <= k; c++ {
				table[i][j][c] = -1
			}
		}
	}

	startCost := 0
	if grid[0][0] > 0 {
		startCost = 1
	}
	if startCost <= k {
		table[0][0][startCost] = grid[0][0]
	}

	for i := 0; i < len(grid); i++ {
		for j := 0; j < len(grid[i]); j++ {
			cost := 0
			if grid[i][j] > 0 {
				cost = 1
			}

			// from top
			if i > 0 {
				for c := 0; c <= k; c++ {
					if cost+c <= k {
						table[i][j][cost+c] = greater(table[i][j][cost+c], table[i-1][j][c]+grid[i][j])
					}
				}
			}

			// from left
			if j > 0 {
				for c := 0; c <= k; c++ {
					if cost+c <= k {
						table[i][j][cost+c] = greater(table[i][j][cost+c], table[i][j-1][c]+grid[i][j])
					}
				}
			}
		}
	}

	best := -1
	target := table[len(grid)-1][len(grid[0])-1]

	for i := 0; i <= k; i++ {
		best = greater(best, target[i])
	}

	return best
}
